src/network/networkd-dhcp4.c

```diff
@@ -944,11 +953,19 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
                                 log_link_notice(link, "Overlong DHCP hostname received, shortened from '%s' to '%s'", dhcpname, hostname);
                 }
 
-                if (hostname) {
-                        r = manager_set_hostname(link->manager, hostname);
+                log_link_debug(link, "Setting DHCP4 hostname '%s'", use_hostname_to_string(link->network->dhcp_use_hostname));
+
+                if (hostname && IN_SET(link->network->dhcp_use_hostname, USE_HOSTNAME_YES, USE_HOSTNAME_BOTH)) {
+                        r = manager_set_hostname(link->manager, hostname, false);
                         if (r < 0)
                                 log_link_error_errno(link, r, "Failed to set transient hostname to '%s': %m", hostname);
                 }
+
+                if (hostname && IN_SET(link->network->dhcp_use_hostname, USE_HOSTNAME_STATIC, USE_HOSTNAME_BOTH)) {
+                        r = manager_set_hostname(link->manager, hostname, true);
+                        if (r < 0)
+                                log_link_error_errno(link, r, "Failed to set static hostname to '%s': %m", hostname);
+                }
```

Comment:
Please move the debug log in `if (hostname)` block. Also, as already commented in the above, `USE_HOSTNAME_BOTH` is meaning less, so how about something like the following:
```
if (hostname) {
    log_link_debug(link, "Setting...");
    r = manager_set_hostname(link->manager, hostname, link->network->dhcp_use_hostname == USE_HOSTNAME_STATIC);
    if (r < 0)
         log_link_error_errno(...);
}
```